pandas/core/indexes/multi.py

```diff
@@ -1143,11 +1143,21 @@ def to_frame(self, index=True):
         """
 
         from pandas import DataFrame
-        result = DataFrame({(name or level):
-                            self._get_level_values(level)
-                            for name, level in
-                            zip(self.names, range(len(self.levels)))},
-                           copy=False)
+        if names is not None:
+            if len(names) != len(self.levels):
+                raise AssertionError("'names' should have same lenght as "
+                                     "number of levels on index")
+            result = DataFrame({(name):
```

Comment:
can you just set the name once here (in the if/else), then construct the frame after, so the construction code is only once